### PR TITLE
Add global keyring management for ECL

### LINKED ISSUE #5613: RFE: ECL configuration commands
Since 3.6.0 release with the introduction of security fix related to signature check, ECL now requires that all users manually import signer key in their local keyring making ECL experience less convenient for both users and administrators.
Since ECL has nothing to do with user keyrings and shouldn't use it for security reasons, it would make sense to provide a mechanism allowing administrators to manage ECL keyring and at the same time ECL configuration by introducing a new set of commands in `singularity config`:

* To enable/disable ECL

```
singularity config ecl enable
singularity config ecl disable
singularity config ecl status
```

* To add/update/list/remove ECL execution group

```
singularity config ecl add --dir /path --mode [whilelist|whitestrict|blacklist] --keys [keylist] <group_name>
singularity config ecl update [--dir /path] [--mode <mode>] --add-keys [keylist] --remove-keys [keylist] <group_name>
singularity config ecl list
singularity config ecl remove <group_name>
```

* To manage ECL keyring

```
singularity config ecl import-key <public-key-path>
singularity config ecl pull-key [--url <url>] <fingerprint>
singularity config ecl list-key
singularity config ecl remove-key <fingerprint>
```

### PR DESCRIPTION
**Pull request by cclerget**
## Description of the Pull Request (PR):

Add a `--global` flag for  `key import/export/list/pull/push/remove` commands to manage the global keyring stored at `SYSCONFDIR/singularity/pgp-public`.

ECL now exclusively use the global keyring.  Global keyring is also used by `singularity verify` with an explicit tag `[GLOBAL]`

### This fixes or addresses the following GitHub issues:

 - Fixes #5613

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by dtrudg** [comment]
Hi @cclerget - as this is a fairly big one I'll block out a chunk of time I can go through it properly Thursday or Friday. I've asked @tri-adam to take a look early next week if he can as he's pretty familiar with the ECL stuff from the sign/verify security rework.

File: cmd/internal/cli/key.go [hunk 1]
```diff
@@ -56,6 +60,16 @@ var keyNewpairBitLengthFlag = cmdline.Flag{
...
 }
 
+// -g|--global
+var keyGlobalPubKeyFlag = cmdline.Flag{
+	ID:           "keyGlobalPubKeyFlag",
+	Value:        &keyGlobalPubKey,
+	DefaultValue: false,
+	Name:         "global",
+	ShortHand:    "g",
+	Usage:        "use global public keys (import/pull/remove are restricted to root user or unprivileged installation only)",
```
Comment (dtrudg) [inline cmd/internal/cli/key.go]:
```suggestion
	Usage:        "manage global public keys (import/pull/remove are restricted to root user or unprivileged installation only)",
```

File: dist/rpm/singularity.spec.in [hunk 2]
```diff
@@ -147,6 +147,7 @@ make DESTDIR=$RPM_BUILD_ROOT install man
 %config(noreplace) %{_sysconfdir}/singularity/*.toml
 %config(noreplace) %{_sysconfdir}/singularity/*.json
 %config(noreplace) %{_sysconfdir}/singularity/*.yaml
+%config(noreplace) %{_sysconfdir}/singularity/pgp-public
```
Comment (dtrudg) [inline dist/rpm/singularity.spec.in]: I'd probably call this `global-pgp-public` just so it includes the word `global` which is throughout the UI now.

File: e2e/ecl/ecl.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+var keyMap = map[string]string{
+	"key1": "0C5B8C9A5FFC44E2A0AC79851CD6FA281D476DD1",
+	"key2": "78F8AD36B0DCB84B707F23853D608DAE21C8CA10",
+}
+
+type ctx struct {
+	env e2e.TestEnv
+}
+
+func (c *ctx) eclConfig(t *testing.T) {
```
Comment (dtrudg) [inline e2e/ecl/ecl.go:27]: In these tests it seems a little odd to be running our own builds etc. when we have test images and keys in `singularity/internal/pkg/syecl/testdata` - I'm just wondering if there is a way to unify the ecl testing a bit so we don't have multiple places to update?

File: mlocal/frags/build_runtime.mk [hunk 4]
```diff
@@ -128,3 +128,15 @@ $(cgroups_config_INSTALL): $(cgroups_config)
 	$(V)install -m 0644 $< $@
 
 INSTALLFILES += $(cgroups_config_INSTALL)
+
+# global keyring
+global_keyring := $(SOURCEDIR)/etc/pgp-public
```
Comment (dtrudg) [inline mlocal/frags/build_runtime.mk]: See other comment - I think global-pgp-public is better to match the 'global' throughout the ui

**Review by dtrudg — CHANGES_REQUESTED** [review]
Looks reasonable to me. I left a few minor comments.

Things I do think are important:

- This code doesn't check that the global keyring is owned by root when ECL is in use. This is inconsistent with the fact that we *do* check the ecl config is owned by root:

```
03:03 PM $ singularity run alpine_latest.sif 
Singularity>

# Still runs with pgp-public owned by me
03:03 PM $ sudo chown dave.dave /usr/local/etc/singularity/pgp-public
03:04 PM $ singularity run alpine_latest.sif 
Singularity> 

# But not if ecl.toml is owned by me
03:04 PM $ sudo chown dave.dave /usr/local/etc/singularity/ecl.toml 
03:04 PM $ singularity run alpine_latest.sif 
FATAL:   /usr/local/etc/singularity/ecl.toml must be owned by root
```

- `docs/content.go` needs to be updated for the key command / subcommands. At present it does not mention the global keyring, and explicitly gives the path to the user keyring etc.

- An issue must be added to the userdocs and admindocs repo, to update for these changes. They need to be blockers for release of 3.7.

File: e2e/ecl/ecl.go (hunk 3 above)
Comment (cclerget) [reply e2e/ecl/ecl.go:27]: Well images in this package are running only on x86 platforms and e2e tests need full execution contrary to syecl package, I'm not sure about e2e tests if busybox is supported for different platforms as tests uses library ... and with docker rate limit coming soon. A clear separation between integration/unit tests and e2e tests regarding test data is good imo and eventually allow to spot different bugs.

Comment (dtrudg) [reply e2e/ecl/ecl.go:27]: okay, fair enough.

**Review by dtrudg — APPROVED** [review]

**Comment by tri-adam** [comment]
The code looks good to me. Only real issue I can see is that the documentation talks about managing global public keys, but the `--secret` flag can be used in combination with `--global`? I'm not sure this is intentional? For example:

```sh
$ singularity key list --global --secret
Private key listing (/usr/local/etc/singularity/pgp-secret)
```

**Comment by cclerget** [comment]
@tri-adam Fixed, checks regarding private keys are done in sypgp package, but `key list` didn't checked errors from calls `PrintPubKeyring` and `PrintPrivKeyring`

**Review by dtrudg — APPROVED** [review]
LGTM for ad's comment

**Pull request merged**